test/hotspot/gtest/concurrent_test_runner.inline.hpp

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (c) 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#ifndef GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
+#define GTEST_CONCURRENT_TEST_RUNNER_INLINE_HPP
+
+#include "threadHelper.inline.hpp"
+
+// This file contains helper classes to run unit tests concurrently in multiple threads.
+
+// Base class for test runnable. Override runUnitTest() to specify what to run.
+class TestRunnable {
+public:
+  virtual void runUnitTest() const = 0;
+};
+
+// This class represents a thread for a unit test.
+class UnitTestThread : public JavaTestThread {
+public:
+  // runnableArg - what to run
+  // doneArg - a semaphore to notify when the thread is done running
+  // testDurationArg - how long to run (in milliseconds)
+  UnitTestThread(TestRunnable* const runnableArg, Semaphore* doneArg, const long testDurationArg) :
+    JavaTestThread(doneArg), runnable(runnableArg), testDuration(testDurationArg) {}
+
+  // from JavaTestThread
+  void main_run() {
+    long stopTime = os::javaTimeMillis() + testDuration;
+    while (os::javaTimeMillis() < stopTime) {
+      runnable->runUnitTest();
+    }
+  }
+private:
+  TestRunnable* const runnable;
+  const long testDuration;
+};
+
+// Helper class for running a given unit test concurrently in multiple threads.
+class ConcurrentTestRunner {
+public:
+  // runnableArg - what to run
+  // nrOfThreadsArg - how many threads to use concurrently
+  // testDurationMillisArg - duration for each test run
+  ConcurrentTestRunner(TestRunnable* const runnableArg, int nrOfThreadsArg, long testDurationMillisArg) :
+    unitTestRunnable(runnableArg),
+    nrOfThreads(nrOfThreadsArg),
+    testDurationMillis(testDurationMillisArg) {}
+
+  virtual ~ConcurrentTestRunner() {}
+
+  void run() {
+    Semaphore done(0);
+
+    std::vector<UnitTestThread*> t;
```

Comment:
Regarding the STL question, I think using std in a limited fashion in the tests seems ok before we start allowing it in the main sources, despite where Misha pointed out it already exists.  It's going to be allowed pretty soon anyway.
One question about this: where does it allocate memory for 't' ?  Are the elements leaked here?

Reply:
On a question: where does it allocate memory for 't' ? Are the elements leaked here?
The vector will allocate and free the memory for the containers (the memory containing pointers to UnitTestThread). 

As for the UnitTestThread objects/instances themselves:
- JavaTestThread. I checked couple of other examples of uses of JavaTestThread (test_symbolTable.cpp and oops/test_markWord.cpp), and neither is calling explicit delete on the objects of this type
- I presume the thread will free its resources once it exits (and we wait for done semafore to ensure this)
- adding code “delete t[i]” results in crash, presuming due to freeing an already released memory

Hence, I believe, no changes needed to this code to free resources.

Reply:
> Regarding the STL question, I think using std in a limited fashion in the tests seems ok before we start allowing it in the main sources, despite where Misha pointed out it already exists. It's going to be allowed pretty soon anyway.

Really? I recently had a private discussion with Kim Barret about this and understood that this is a contentious point. I was hoping we would have a public discussion about this before deciding on this. My fears are increased build times (which seem to get worse and worse) and stability- and compiler issues. I am not completely against it, but I'm a bit of a burned child wrt to STL.

Reply:
> One question about this: where does it allocate memory for 't' ? Are the elements leaked here?

Backing buffer for vector lives in C-heap. Its a dynamically growing array, basically like our GrowableArray. Only its not under our control and we cannot easily debug it if something goes wrong (take a look at the STL sources, they are not easy to understand). And it sidesteps our own os::malloc, so we cannot account it, and our guards won't work.

Personally I would like to keep the test harness as simple stupid as possible, to avoid interfering with the actual test. Especially with tests which test memory allocation. I know that is a vague reason, but I also cannot find a bit advantage in using std::vector, compared with a simple calloced array. We dont need the dynamically-growing part here, we know right upfront how many threads we start, so its really a fixed sized array.

Reply:
Okay so you're right that we can't account in NMT and with our malloc guards for the memory that is allocated with std::vector, so I'm convinced that Misha should change this.  Hopefully we can figure out how to make std::vector work in the future, because Misha's 2 lines for it look quite nice.
Do you think using std:: would increase build times? My suspicion is that the Access barrier templates are to blame, but building hotspot is a fraction of building the whole jdk, which does seem to take a long time.

Reply:
> Okay so you're right that we can't account in NMT and with our malloc guards for the memory that is allocated with std::vector, so I'm convinced that Misha should change this. Hopefully we can figure out how to make std::vector work in the future, because Misha's 2 lines for it look quite nice.
> Do you think using std:: would increase build times? My suspicion is that the Access barrier templates are to blame, but building hotspot is a fraction of building the whole jdk, which does seem to take a long time.

Well, yesterday I saw to my annoyance that upgrading from g++-7 to g++-9 increased jdk build time by about 30-40%, so I today I am very aware of build times :-/ But you are right of course, its just the hotspot, so maybe its not such a big deal.

I am a bit apprehensive about introducing STL (had to port the STL to outlier platforms in the past and god that was terrible) but mainly I'd like to have a public discussion before starting to use it. And if the majority thinks its okay then I am okay with it too oc. I have been wrong before. Kim said STL quality is very high nowadays.

Reply:
Thanks for your feedback and discussion. I will refrain from using STL, and use malloc to allocate the array. I presume I should use os::malloc() in this case - please correct me if I am wrong.

Reply:
re: "there are some configuration and build kludges that were required to make that work." - Thinking about it more, it might be that those kludges were needed even without the jfr uses of the Standard Library.  I think that may have been needed because of uses of the Standard Library by the gtest framework itself.  The jfr tests then took advantage of that support.